prow/plugins/commands/generics/generics.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generics
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"k8s.io/test-infra/prow/errorutil"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/plugins/commands"
+	"k8s.io/test-infra/prow/plugins/commands/handlers"
+	"k8s.io/test-infra/prow/plugins/commands/policies"
+)
+
+const reFormatAddRemove = `(?mi)^/(remove-)?%s(?: +(.*?))?\s*$`
+
+// AddRemoveCommands take the form `/[remove-]<command> [<args>]`.
+func AddRemoveCommand(command string, issueType commands.IssueType, add, remove commands.MatchHandler) *commands.Command {
+	handler := func(ctx *commands.Context) error {
+		var addMatches, removeMatches [][]string
+		for _, match := range ctx.Matches {
+			if len(match) != 3 {
+				return fmt.Errorf("expected 3 regexp match groups, but got %d", len(match))
+			}
+			if strings.ToLower(match[1]) == "remove-" {
+				removeMatches = append(removeMatches, match)
+			} else {
+				addMatches = append(addMatches, match)
+			}
+		}
+		var err error
+		if len(removeMatches) > 0 {
+			ctx.Matches = removeMatches
+			err = remove(ctx)
+		}
+		if len(addMatches) > 0 {
+			ctx.Matches = addMatches
+			err = errorutil.NewAggregate(err, add(ctx))
```

Comment:
Seems like `add(ctx, addMatches)`  would be easier to understand for the `/foo, /remove-foo` case. 

Although I do not understand the point of `ctc.Matches = addMatches` given that EnsureLabel doesn't seem to use it.

Reply:
`add(ctx, addMatches)` would mean that `add` is not a MatchHandler which would break the compositional structure. I agree that would make sense in this specific case, but that is only because we are filtering the matches here. MatchHandlers already receive the matches as part of the context so the parameter is just being passed there.
I could move `Matches` out of `Context` and change the `MatchHandler` type to `func(ctx *Context, matches [][]string) error` if that is preferred.

`ctx.Matches = addMatches` doesn't have any affect when `add` is `EnsureLabel`, but that isn't necessarily the case. `add` is an arbitrary MatchHandler so even though we happen to be using a MatchHandler that doesn't consume the matches, other MatchHandlers could be provided in place of `EnsureLabel` and they may consume the matches.

Reply:
> I could move Matches out of Context and change the MatchHandler type to func(ctx *Context, matches [][]string) error if that is preferred.

Makes sense that others might want to use this. However if the initial use of `MatchHandler` is awkward that suggests we have the wrong abstraction. So I would suggest changing something here.